packages/IntegrationTest/windows/integrationtest/TestHostHarness.cpp

```diff
@@ -0,0 +1,195 @@
+      co_return;
+
+    case ListenResult::OtherError:
+      ShowJSError("The test harness ran into an unexpected error trying to listen for commands.");
+      co_return;
+
+    default:
+      ShowJSError("Unexpected ListenResult from TestCommandListener.");
+      co_return;
+  }
+}
+
+winrt::fire_and_forget TestHostHarness::OnTestCommand(TestCommand command, TestCommandResponse response) noexcept {
+  // Keep ourselves alive while we have a test command
+  auto strongThis = get_strong();
+
+  switch (command.Id) {
+    case TestCommandId::RunTestComponent: {
+      auto componentName = command.payload.GetString();
+      m_rootView.ComponentName(componentName);
+
+      // We might see multiple JS events for exceptions, failing tests, etc.
+      // Flush everything pending from the JS thread to UI thread before
+      // continuing to ensure all events from the past command are handled.
+      co_await FlushJSQueue();
+
+      m_pendingResponse = std::move(response);
+      m_timeoutAction = TimeoutIfNoResponse();
+      break;
+    }
+
+    case TestCommandId::GoToComponent: {
+      auto componentName = command.payload.GetString();
+      m_rootView.ComponentName(componentName);
+      co_await FlushJSQueue();
+      response.Okay();
+      break;
+    }
+
+    default: {
+      ShowJSError("Unexpected command ID from test runner");
+      response.UnknownError("Unexpected command ID from test runner");
+      break;
+    }
+  }
+}
+
+void TestHostHarness::OnTestCompleted() noexcept {
+  OnTestPassed(true);
+}
+
+void TestHostHarness::OnTestPassed(bool passed) noexcept {
+  if (m_pendingResponse) {
+    m_pendingResponse->TestPassed(passed);
+    CompletePendingResponse();
+  }
+}
+
+IAsyncAction TestHostHarness::TimeoutIfNoResponse() noexcept {
+  constexpr auto CommandTimeout = 20s;
+
+  auto weakThis = get_weak();
+  co_await CommandTimeout;
+
+  auto cancellationToken = co_await winrt::get_cancellation_token();
+  if (auto strongThis = weakThis.get() && m_pendingResponse && !cancellationToken()) {
+    m_pendingResponse->Timeout();
+    CompletePendingResponse();
+  }
+}
+
+IAsyncAction TestHostHarness::FlushJSQueue() noexcept {
+  winrt::handle signal(CreateEvent(nullptr, false, false, nullptr));
+
+  m_context.JSDispatcher().Post([&signal, uiDispatcher{m_context.UIDispatcher()} ]() noexcept {
+    uiDispatcher.Post([&signal]() noexcept { SetEvent(signal.get()); });
+  });
+
+  co_await winrt::resume_on_signal(signal.get());
+}
+
+void TestHostHarness::ShowJSError(std::string_view err) noexcept {
+  m_context.CallJSFunction(L"RCTLog", L"logToConsole", "error", err);
+}
+
+void TestHostHarness::CompletePendingResponse() noexcept {
+  if (m_pendingResponse) {
+    m_pendingResponse.reset();
+  }
+
+  if (m_timeoutAction) {
+    m_timeoutAction.Cancel();
+  }
+}
+
+void TestHostHarnessRedboxHandler::ShowNewError(IRedBoxErrorInfo info, RedBoxErrorType /*type*/) {
```

Comment:
We should care about type. Fatal exceptions signal an end to the test  where non-fatal errors don't.